### PR TITLE
Add context reset feature to IoT.js

### PR DESCRIPTION
**Pull request by DanielBallaSZTE**
IoT.js now also properly handles the context reset command sent by IoT.JS Code.

IoT.js-DCO-1.0-Signed-off-by: Daniel Balla dballa@inf.u-szeged.hu

### CODE REVIEW

**Comment by glistening** [comment]
@DanielBallaSZTE Are you adding feature that enables restarting iotjs in debugger? If so, does context reset work well in tizenrt or/and nuttx? I am not sure iotjs clean up for several kind of termination. In fact, there is an issue #1245 related to termination.

**Comment by DanielBallaSZTE** [comment]
@glistening Yes, this feature would make it possible to restart IoT.js from the debugger client. On nuttx 
 the debugger wasn't working anyways with IoT.js. We can test next week if it works on TizenRT however, and report the results.

**Comment by glistening** [comment]
@DanielBallaSZTE I would like to land this PR after testing is done on tizenrt. Please let me know when you completed test.

**Comment by robertsipka** [comment]
There is an issue where several error appears in the build of the iotjs project besides TizenRT/Nuttx with enabled jerry-debugger on artik053/stm32f4 device.  It is also occurs on the master branch, I will investigate it.

**Comment by robertsipka** [comment]
I've successfully  tried out this PR on artik053, after I got over of the build issue.
It sends the context reset to the server, but after that the server does not responds with a `waiting-for-a-source` status code. At least the client from the `iotjscode` does not receive that, it stuck in the running mode.

File: src/iotjs.c [hunk 1]
```diff
@@ -178,6 +178,10 @@ static void iotjs_uv_walk_to_close_callback(uv_handle_t* handle, void* arg) {
   iotjs_handlewrap_close(handle_wrap, NULL);
 }
 
+static jerry_value_t dummy_wait_for_client_source_cb() {
+  return jerry_create_undefined();
+}
+
```
Comment (robertsipka) [inline src/iotjs.c]: Please remove this extra line.

**Comment by daeyeon** [comment]
@DanielBallaSZTE @robertsipka I also confirmed that this works on rpi3 with iotjscode. Can you kindly update the status related to artik053? We're waiting for this commit landing. :-)

**Comment by robertsipka** [comment]
I'am planning to debug this behavior on artik053, but I have not had time yet. I will raise it's priority in my TODO list.

**Comment by robertsipka** [comment]
I don't know what was the problem last time (maybe the version of the IDE), but it works now. I've applied this patch to the master and checkout the latest `iotjscode` to try with it. The context reset works properly on artik053. LGTM (informally)

**Comment by DanielBallaSZTE** [comment]
@robertsipka Thanks for testing it on artik053 as well.
@glistening @daeyeon Are there any other requests on this pull request?

**Comment by daeyeon** [comment]
@robertsipka While building iotjs with `--jerry-debugger`, I observed the below error. Can you share your steps to build it?

```shell
[ 61%] Building C object jerry-core/CMakeFiles/jerry-core.dir/ecma/operations/ecma-jobqueue.c.obj
/home/daeyeon/Code/build/iotjs-tizenrt/iotjs/deps/jerry/jerry-core/debugger/debugger-ws.c:22:23: fatal error: arpa/inet.h: No such file or directory
 #include <arpa/inet.h>
                       ^
```

**Comment by robertsipka** [comment]
@daeyeon :Yes, as I mentioned earlier https://github.com/Samsung/iotjs/pull/1274#issuecomment-339040314, there are some required steps to build it properly:
1. To solve the error you mentioned, it is required to add the include directory of the header file, so I use this command: 
`make IOTJS_ROOT_DIR=../../iotjs IOTJS_BUILD_OPTION="--jerry-debugger --compile-flag=-I/home/sipka/SOW11/TizenRT/os/include/"`
2. Furthermore, because the CMakeLists.txt of JerryScript (a dependency of IoT.js) sets the conversion compile warning as an error with -Werror=conversion flag, I've got several conversion errors both from TizenRT and JerryScript source (with the compiler that I used for the build).
To solve this, I've uploaded a pull request in TizenRT, and suggested to reopen a pull request in JerryScript (which resolve this problem), but you can temporarily remove the conversion flag from this line: https://github.com/jerryscript-project/jerryscript/blob/7692aa9d660a3ecf4d4772bdb8b426ac85b6a98d/CMakeLists.txt#L208
3. There are some implicit declarations in th  `debbugger-ws.c`, because those declarations are probably available elsewhere within the libc that TizenRT use. You need to include `<sys/socket.h>` in here:  
https://github.com/jerryscript-project/jerryscript/blob/bdcd2d81797e96571457e0f49c97fc0245d08acf/jerry-core/debugger/debugger-ws.c#L22
It seems that we need to create a target specific solution for this problem.

**Comment by daeyeon** [comment]
@robertsipka Thank you for the details.

I observed this feature is blocked when user set any break point. Can you check the below steps?

i) Click `Run` button on `Source sending` to send a source.
ii) Set a breakpoint.
iii) Click `Continue & Stop` to reach the above breakpoint.
iv) Click `Continue & Stop` to go through the source.
v) `The engine is waiting for a source.` toast is expected, but not popped up.
    `Run` on `Source sending` remains as 'disabled'.

File: src/iotjs.c [hunk 2]
```diff
@@ -216,15 +219,37 @@ int iotjs_entry(int argc, char** argv) {
...
+    context_reset = iotjs_environment_config(env)->debugger->context_reset;
+  }
   // Release environment.
   iotjs_environment_release();
 
   // Release debug print setting.
   release_debug_settings();
 
-  return ret_code;
+  if (context_reset) {
```
Comment (daeyeon) [inline src/iotjs.c]: I suggest removing `else` block.

``` javascript
if (context_reset) {
   return iotjs_entry(argc, argv);
}
return ret_code;
```

File: src/iotjs.c [hunk 3]
```diff
@@ -216,15 +219,37 @@ int iotjs_entry(int argc, char** argv) {
...
+      iotjs_environment_config(env)->debugger->context_reset = true;
+    }
+    jerry_release_value(res);
+  }
+
   // Release JerryScript engine.
   iotjs_jerry_release(env);
 
-terminate:
+terminate:;
```
Comment (daeyeon) [inline src/iotjs.c]: Leaving `;` here looks strange. I also failed to notice it at my first glance. Do we surely need this?

Comment (DanielBallaSZTE) [reply src/iotjs.c]: You must have a statement after a goto label, and variable declaration isn't one, so ; counts as an empty statement.

**Comment by robertsipka** [comment]
@daeyeon:
I've tried your steps using the iotjscode (on master, hash: 1fe29dea). It worked for me on artik053 and x86 as well.  After the second  click on `Continue & Stop` the debugger went through the source, context reset sent automatically and the info message was popped up with the `Waiting for a source` message.
Run on Source sending is enabled, the `Selected file(s)` list contains the latest js.

**Comment by DanielBallaSZTE** [comment]
@daeyeon 
I can only confirm what @robertsipka said, tested it multiple times, works for me (x86).

**Comment by daeyeon** [comment]
@robertsipka @DanielBallaSZTE Thanks for the swift response. I maybe did something wrong. I will check that again. Plus, sorry for the belated review of this commit. Please fix the conflict. LGTM except it.

**Comment by DanielBallaSZTE** [comment]
@daeyeon There's some issue with the patch at the moment on latest iotjs master, I'm working on a fix right now.

**Comment by DanielBallaSZTE** [comment]
@daeyeon Actually @galpeter found a solution for the issue in https://github.com/Samsung/iotjs/pull/1321
I suggest merging his PR first, so mine works properly.

**Comment by yichoi** [comment]
LGTM

**Pull request merged**